File: evaluator/metrics.py
import numpy as np
from typing import Dict, List, Tuple, Any, Optional, Union

# Standard COCO 10-step IoU thresholds: [0.50:0.95:0.05]
COCO_IOU_THRESHOLDS: List[float] = [
    0.50, 0.55, 0.60, 0.65, 0.70, 0.75, 0.80, 0.85, 0.90, 0.95
]

def parse_iou_thresholds(arg: Union[str, List[float], Tuple[float, ...], None]) -> List[float]:
    """Parses IoU thresholds from a CLI string, list of floats, or alias.
    
    Supports:
        - None / 'coco' / 'default' / 'standard': Standard COCO 10-step thresholds [0.50:0.95:0.05]
        - Range string: '0.5:0.95:0.05' or '0.5:0.95'
        - Comma-separated floats: '0.5' or '0.3,0.5,0.75'
        - List/Tuple of floats: [0.5, 0.75]
    """
    if arg is None:
        return list(COCO_IOU_THRESHOLDS)
        
    if isinstance(arg, (list, tuple)):
        thresholds = [float(x) for x in arg]
    elif isinstance(arg, str):
        clean_arg = arg.strip().lower()
        if clean_arg in ("coco", "default", "standard"):
            return list(COCO_IOU_THRESHOLDS)
            
        if ":" in clean_arg:
            parts = [float(p.strip()) for p in clean_arg.split(":")]
            if len(parts) == 2:
                start, end = parts
                step = 0.05
            elif len(parts) == 3:
                start, end, step = parts
            else:
                raise ValueError(f"Invalid range format '{arg}'. Use 'start:stop' or 'start:stop:step'.")
            if step <= 0:
                raise ValueError(f"Step in range '{arg}' must be positive.")
            # Use np.arange with buffer to ensure upper boundary inclusion
            vals = np.arange(start, end + (step / 2.0), step)
            thresholds = [round(float(v), 4) for v in vals]
        else:
            thresholds = [float(x.strip()) for x in clean_arg.split(",") if x.strip()]
    else:
        raise TypeError(f"Expected str, list of floats, or None for iou_thresholds, got {type(arg).__name__}")
        
    if not thresholds:
        raise ValueError("IoU thresholds cannot be empty.")
        
    for th in thresholds:
        if not (0.0 <= th <= 1.0):
            raise ValueError(f"IoU threshold {th} must be between 0.0 and 1.0.")
            
    # Return sorted unique list
    return sorted(list(set(round(th, 4) for th in thresholds)))

def compute_iou(box1: List[float], box2: List[float]) -> float:
    """Computes Intersection over Union (IoU) of two bounding boxes.
    
    Boxes are in format [ymin, xmin, ymax, xmax].
    """
    ymin1, xmin1, ymax1, xmax1 = box1
    ymin2, xmin2, ymax2, xmax2 = box2
    
    y_min = max(ymin1, ymin2)
    x_min = max(xmin1, xmin2)
    y_max = min(ymax1, ymax2)
    x_max = min(xmax1, xmax2)
    
    if y_max <= y_min or x_max <= x_min:
        return 0.0
        
    intersection = (y_max - y_min) * (x_max - x_min)
    area1 = (ymax1 - ymin1) * (xmax1 - xmin1)
    area2 = (ymax2 - ymin2) * (xmax2 - xmin2)
    
    union = area1 + area2 - intersection
    if union <= 0.0:
        return 0.0
        
    return intersection / union

class DetectionEvaluator:
    """Computes Object Detection metrics (mAP, Precision, Recall) at multiple IoU thresholds,
    supporting standard COCO benchmark evaluation (mAP@[.50:.95], mAP@.50, mAP@.75).
    """
    
    def __init__(
        self, 
        label_map: Optional[Dict[str, str]] = None, 
        iou_thresholds: Optional[Union[List[float], Tuple[float, ...], str]] = None,
        ap_method: str = "all_points"
    ) -> None:
        self.label_map = label_map or {}
        self.iou_thresholds = parse_iou_thresholds(iou_thresholds)
        self.ap_method = ap_method
        self.all_predictions: List[Dict[str, Any]] = []
        self.all_ground_truths: List[Dict[str, Any]] = []
        self.image_counter = 0

    def update(self, predictions: List[Dict[str, Any]], ground_truths: List[Dict[str, Any]]) -> None:
        """Adds predictions and ground truths for a single image to the evaluator state.
        
        Args:
            predictions: List of dicts like: [{'bbox': [ymin, xmin, ymax, xmax], 'label': str, 'score': float}]
            ground_truths: List of dicts like: [{'bbox': [ymin, xmin, ymax, xmax], 'label': str}]
        """
        img_idx = self.image_counter
        self.image_counter += 1
        
        for p in predictions:
            # Map predictions to standard vocabulary
            raw_label = p["label"]
            mapped_label = self.label_map.get(raw_label, raw_label)
            self.all_predictions.append({
                "image_id": img_idx,
                "bbox": p["bbox"],
                "label": mapped_label,
                "score": p.get("score", 1.0),
                "orig_label": raw_label
            })
            
        for gt in ground_truths:
            self.all_ground_truths.append({
                "image_id": img_idx,
                "bbox": gt["bbox"],
                "label": gt["label"]
            })

    def _calculate_ap_for_class(
        self, 
        cls_preds: List[Dict[str, Any]], 
        cls_gts: List[Dict[str, Any]], 
        iou_threshold: float
    ) -> Tuple[float, float, float]:
        """Calculates AP, Precision, and Recall for a single class at a given IoU threshold."""
        total_gts = len(cls_gts)
        if total_gts == 0:
            return 0.0, 0.0, 0.0
            
        if len(cls_preds) == 0:
            return 0.0, 0.0, 0.0
            
        # Sort predictions by score descending
        cls_preds = sorted(cls_preds, key=lambda x: x["score"], reverse=True)
        
        # Group ground truths by image ID
        gt_by_img: Dict[int, List[Tuple[int, List[float]]]] = {}
        for idx, gt in enumerate(cls_gts):
            img_id = gt["image_id"]
            if img_id not in gt_by_img:
                gt_by_img[img_id] = []
            gt_by_img[img_id].append((idx, gt["bbox"]))
            
        # Track matched ground truths
        matched_gt_indices = set()
        
        tp = np.zeros(len(cls_preds))
        fp = np.zeros(len(cls_preds))
        
        for i, pred in enumerate(cls_preds):
            img_id = pred["image_id"]
            pred_box = pred["bbox"]
            
            best_iou = -1.0
            best_gt_idx = -1
            
            if img_id in gt_by_img:
                for gt_idx, gt_box in gt_by_img[img_id]:
                    iou = compute_iou(pred_box, gt_box)
                    if iou > best_iou:
                        best_iou = iou
                        best_gt_idx = gt_idx
                        
            if best_iou >= iou_threshold:
                if best_gt_idx not in matched_gt_indices:
                    tp[i] = 1.0
                    matched_gt_indices.add(best_gt_idx)
                else:
                    fp[i] = 1.0  # Duplicate prediction
            else:
                fp[i] = 1.0
                
        cumulative_tp = np.cumsum(tp)
        cumulative_fp = np.cumsum(fp)
        
        recalls = cumulative_tp / total_gts
        precisions = cumulative_tp / (cumulative_tp + cumulative_fp + 1e-16)
        
        # Compute AP using selected method
        if self.ap_method == "coco_101":
            # 101-point COCO interpolation at recall points 0.00:0.01:1.00
            rec_thresholds = np.linspace(0.0, 1.0, 101)
            precisions_at_rec = []
            for r in rec_thresholds:
                mask = recalls >= r
                if np.any(mask):
                    precisions_at_rec.append(np.max(precisions[mask]))
                else:
                    precisions_at_rec.append(0.0)
            ap = float(np.mean(precisions_at_rec))
        else:
            # Continuous all-points interpolation (PASCAL VOC 2012+ / continuous AUC)
            mrec = np.concatenate(([0.0], recalls, [1.0]))
            mpre = np.concatenate(([0.0], precisions, [0.0]))
            
            # Compute the precision envelope
            for j in range(len(mpre) - 2, -1, -1):
                mpre[j] = max(mpre[j], mpre[j + 1])
                
            # Integrate area under curve
            split_indices = np.where(mrec[1:] != mrec[:-1])[0]
            ap = float(np.sum((mrec[split_indices + 1] - mrec[split_indices]) * mpre[split_indices + 1]))
        
        # Final precision and recall values at the end of prediction list
        final_precision = float(precisions[-1]) if len(precisions) > 0 else 0.0
        final_recall = float(recalls[-1]) if len(recalls) > 0 else 0.0
        
        return ap, final_precision, final_recall

    def compute_metrics(self) -> Dict[str, Any]:
        """Calculates evaluation metrics over all collected state."""
        # Get unique classes present in ground truths or predictions
        all_classes = set(gt["label"] for gt in self.all_ground_truths) | set(pred["label"] for pred in self.all_predictions)
        all_classes = sorted(list(all_classes))
        
        report: Dict[str, Any] = {
            "per_iou": {},
            "mAP_50_95": 0.0,
            "mAP_50": 0.0,
            "mAP_75": 0.0,
            "mAP_mean": 0.0,
            "is_coco_standard": False,
            "image_count": self.image_counter,
            "total_predictions": len(self.all_predictions),
            "total_ground_truths": len(self.all_ground_truths),
            "per_class_summary": {}
        }
        
        map_values = []
        
        for iou_th in self.iou_thresholds:
            iou_key = f"iou_{iou_th:.2f}"
            iou_report = {}
            class_aps = []
            class_precisions = []
            class_recalls = []
            
            for cls in all_classes:
                cls_preds = [p for p in self.all_predictions if p["label"] == cls]
                cls_gts = [g for g in self.all_ground_truths if g["label"] == cls]
                
                ap, precision, recall = self._calculate_ap_for_class(cls_preds, cls_gts, iou_th)
                
                iou_report[cls] = {
                    "AP": ap,
                    "precision": precision,
                    "recall": recall,
                    "num_ground_truth": len(cls_gts),
                    "num_predictions": len(cls_preds)
                }
                
                if len(cls_gts) > 0:
                    class_aps.append(ap)
                    class_precisions.append(precision)
                    class_recalls.append(recall)
                    
            mAP = float(np.mean(class_aps)) if len(class_aps) > 0 else 0.0
            mean_precision = float(np.mean(class_precisions)) if len(class_precisions) > 0 else 0.0
            mean_recall = float(np.mean(class_recalls)) if len(class_recalls) > 0 else 0.0
            
            report["per_iou"][iou_key] = {
                "mAP": mAP,
                "mean_precision": mean_precision,
                "mean_recall": mean_recall,
                "class_metrics": iou_report
            }
            
            map_values.append(mAP)
            if abs(iou_th - 0.50) < 1e-4:
                report["mAP_50"] = mAP
            if abs(iou_th - 0.75) < 1e-4:
                report["mAP_75"] = mAP
                
        report["mAP_mean"] = float(np.mean(map_values)) if len(map_values) > 0 else 0.0

        # Check if all 10 standard COCO thresholds (0.50:0.05:0.95) are present
        coco_keys = [f"iou_{th:.2f}" for th in COCO_IOU_THRESHOLDS]
        is_coco_std = all(k in report["per_iou"] for k in coco_keys)
        report["is_coco_standard"] = is_coco_std

        if is_coco_std:
            coco_maps = [report["per_iou"][k]["mAP"] for k in coco_keys]
            report["mAP_50_95"] = float(np.mean(coco_maps))
        else:
            report["mAP_50_95"] = report["mAP_mean"]

        # Build per-class summary across thresholds
        for cls in all_classes:
            cls_aps_across_ious = [
                report["per_iou"][f"iou_{th:.2f}"]["class_metrics"][cls]["AP"] 
                for th in self.iou_thresholds
            ]
            
            ap_50 = report["per_iou"]["iou_0.50"]["class_metrics"][cls]["AP"] if "iou_0.50" in report["per_iou"] else 0.0
            ap_75 = report["per_iou"]["iou_0.75"]["class_metrics"][cls]["AP"] if "iou_0.75" in report["per_iou"] else 0.0
            prec_50 = report["per_iou"]["iou_0.50"]["class_metrics"][cls]["precision"] if "iou_0.50" in report["per_iou"] else 0.0
            rec_50 = report["per_iou"]["iou_0.50"]["class_metrics"][cls]["recall"] if "iou_0.50" in report["per_iou"] else 0.0

            if is_coco_std:
                coco_cls_aps = [report["per_iou"][k]["class_metrics"][cls]["AP"] for k in coco_keys]
                ap_coco = float(np.mean(coco_cls_aps))
            else:
                ap_coco = float(np.mean(cls_aps_across_ious)) if cls_aps_across_ious else 0.0

            num_gts = len([g for g in self.all_ground_truths if g["label"] == cls])
            num_preds = len([p for p in self.all_predictions if p["label"] == cls])

            report["per_class_summary"][cls] = {
                "AP_50_95": ap_coco,
                "AP_50": ap_50,
                "AP_75": ap_75,
                "AP_mean": float(np.mean(cls_aps_across_ious)) if cls_aps_across_ious else 0.0,
                "precision_50": prec_50,
                "recall_50": rec_50,
                "num_ground_truth": num_gts,
                "num_predictions": num_preds
            }
        
        return report

File: src/evaluator/metrics.ts
import {
  BoundingBox,
  GroundTruthBox,
  PredictionBox,
  EvaluationReport,
  IoUReport,
  ClassMetric,
  ClassSummary,
} from '../types';

export const COCO_IOU_THRESHOLDS: number[] = [
  0.50, 0.55, 0.60, 0.65, 0.70, 0.75, 0.80, 0.85, 0.90, 0.95
];

export function computeIoU(box1: BoundingBox, box2: BoundingBox): number {
  const [ymin1, xmin1, ymax1, xmax1] = box1;
  const [ymin2, xmin2, ymax2, xmax2] = box2;

  const yMin = Math.max(ymin1, ymin2);
  const xMin = Math.max(xmin1, xmin2);
  const yMax = Math.min(ymax1, ymax2);
  const xMax = Math.min(xmax1, xmax2);

  if (yMax <= yMin || xMax <= xMin) {
    return 0.0;
  }

  const intersection = (yMax - yMin) * (xMax - xMin);
  const area1 = (ymax1 - ymin1) * (xmax1 - xmin1);
  const area2 = (ymax2 - ymin2) * (xmax2 - xmin2);

  const union = area1 + area2 - intersection;
  if (union <= 0.0) {
    return 0.0;
  }

  return intersection / union;
}

export function parseIoUThresholds(
  arg?: string | number[] | null
): number[] {
  if (!arg || (typeof arg === 'string' && ['coco', 'default', 'standard'].includes(arg.trim().toLowerCase()))) {
    return [...COCO_IOU_THRESHOLDS];
  }

  if (Array.isArray(arg)) {
    return Array.from(new Set(arg.map((x) => Math.round(x * 10000) / 10000))).sort((a, b) => a - b);
  }

  const clean = arg.trim().toLowerCase();
  if (clean.includes(':')) {
    const parts = clean.split(':').map((p) => parseFloat(p.trim()));
    let start = parts[0];
    let end = parts[1];
    let step = parts.length === 3 ? parts[2] : 0.05;

    if (step <= 0) step = 0.05;
    const res: number[] = [];
    for (let v = start; v <= end + step / 2.0; v += step) {
      res.push(Math.round(v * 10000) / 10000);
    }
    return Array.from(new Set(res)).sort((a, b) => a - b);
  }

  const values = clean
    .split(',')
    .map((s) => parseFloat(s.trim()))
    .filter((n) => !isNaN(n) && n >= 0 && n <= 1);

  return values.length > 0 ? Array.from(new Set(values)).sort((a, b) => a - b) : [...COCO_IOU_THRESHOLDS];
}

interface StoredPred {
  imageId: number;
  bbox: BoundingBox;
  label: string;
  score: number;
  origLabel: string;
}

interface StoredGT {
  imageId: number;
  bbox: BoundingBox;
  label: string;
}

export class DetectionEvaluator {
  private labelMap: Record<string, string>;
  private iouThresholds: number[];
  private apMethod: 'all_points' | 'coco_101';
  private confThreshold: number;
  private allPredictions: StoredPred[] = [];
  private allGroundTruths: StoredGT[] = [];
  private imageCounter: number = 0;

  constructor(options?: {
    labelMap?: Record<string, string>;
    iouThresholds?: number[] | string;
    apMethod?: 'all_points' | 'coco_101';
    confThreshold?: number;
  }) {
    this.labelMap = options?.labelMap || {};
    this.iouThresholds = parseIoUThresholds(options?.iouThresholds);
    this.apMethod = options?.apMethod || 'all_points';
    this.confThreshold = options?.confThreshold ?? 0.0;
  }

  public update(predictions: PredictionBox[], groundTruths: GroundTruthBox[]): void {
    const imgIdx = this.imageCounter++;

    for (const p of predictions) {
      const score = p.score ?? 1.0;
      if (score < this.confThreshold) continue;

      const rawLabel = p.label;
      const mappedLabel = this.labelMap[rawLabel] || rawLabel;

      this.allPredictions.push({
        imageId: imgIdx,
        bbox: p.bbox,
        label: mappedLabel,
        score,
        origLabel: rawLabel,
      });
    }

    for (const gt of groundTruths) {
      this.allGroundTruths.push({
        imageId: imgIdx,
        bbox: gt.bbox,
        label: gt.label,
      });
    }
  }

  private calculateAPForClass(
    clsPreds: StoredPred[],
    clsGts: StoredGT[],
    iouThreshold: number
  ): [number, number, number] {
    const totalGts = clsGts.length;
    if (totalGts === 0 || clsPreds.length === 0) {
      return [0.0, 0.0, 0.0];
    }

    // Sort predictions by confidence descending
    const sortedPreds = [...clsPreds].sort((a, b) => b.score - a.score);

    // Group ground truths by image ID
    const gtByImg = new Map<number, { idx: number; bbox: BoundingBox }[]>();
    for (let idx = 0; idx < clsGts.length; idx++) {
      const gt = clsGts[idx];
      if (!gtByImg.has(gt.imageId)) {
        gtByImg.set(gt.imageId, []);
      }
      gtByImg.get(gt.imageId)!.push({ idx, bbox: gt.bbox });
    }

    const matchedGtIndices = new Set<number>();
    const tp: number[] = new Array(sortedPreds.length).fill(0);
    const fp: number[] = new Array(sortedPreds.length).fill(0);

    for (let i = 0; i < sortedPreds.length; i++) {
      const pred = sortedPreds[i];
      let bestIoU = -1.0;
      let bestGtIdx = -1;

      const imgGts = gtByImg.get(pred.imageId);
      if (imgGts) {
        for (const { idx, bbox } of imgGts) {
          const iou = computeIoU(pred.bbox, bbox);
          if (iou > bestIoU) {
            bestIoU = iou;
            bestGtIdx = idx;
          }
        }
      }

      if (bestIoU >= iouThreshold) {
        if (!matchedGtIndices.has(bestGtIdx)) {
          tp[i] = 1.0;
          matchedGtIndices.add(bestGtIdx);
        } else {
          fp[i] = 1.0; // Duplicate detection
        }
      } else {
        fp[i] = 1.0;
      }
    }

    // Cumulative sums
    let cumTp = 0;
    let cumFp = 0;
    const recalls: number[] = [];
    const precisions: number[] = [];

    for (let i = 0; i < sortedPreds.length; i++) {
      cumTp += tp[i];
      cumFp += fp[i];
      recalls.push(cumTp / totalGts);
      precisions.push(cumTp / (cumTp + cumFp + 1e-16));
    }

    let ap = 0.0;
    if (this.apMethod === 'coco_101') {
      // 101-point interpolation at points 0.00:0.01:1.00
      const precsAtRec: number[] = [];
      for (let rIdx = 0; rIdx <= 100; rIdx++) {
        const r = rIdx / 100.0;
        let maxPrec = 0.0;
        for (let i = 0; i < recalls.length; i++) {
          if (recalls[i] >= r && precisions[i] > maxPrec) {
            maxPrec = precisions[i];
          }
        }
        precsAtRec.push(maxPrec);
      }
      ap = precsAtRec.reduce((acc, v) => acc + v, 0) / 101.0;
    } else {
      // Continuous all-points interpolation (PASCAL VOC 2012+ / AUC)
      const mrec = [0.0, ...recalls, 1.0];
      const mpre = [0.0, ...precisions, 0.0];

      // Precision envelope
      for (let j = mpre.length - 2; j >= 0; j--) {
        mpre[j] = Math.max(mpre[j], mpre[j + 1]);
      }

      // Integrate
      for (let i = 0; i < mrec.length - 1; i++) {
        if (mrec[i + 1] !== mrec[i]) {
          ap += (mrec[i + 1] - mrec[i]) * mpre[i + 1];
        }
      }
    }

    const finalPrecision = precisions.length > 0 ? precisions[precisions.length - 1] : 0.0;
    const finalRecall = recalls.length > 0 ? recalls[recalls.length - 1] : 0.0;

    return [ap, finalPrecision, finalRecall];
  }

  public computeMetrics(metadata?: Record<string, any>): EvaluationReport {
    const classSet = new Set<string>();
    for (const g of this.allGroundTruths) classSet.add(g.label);
    for (const p of this.allPredictions) classSet.add(p.label);
    const allClasses = Array.from(classSet).sort();

    const report: EvaluationReport = {
      metadata: {
        model_name: metadata?.model_name || 'VLM-Model',
        dataset_path: metadata?.dataset_path || 'datasets/sample',
        dataset_split: metadata?.dataset_split || 'all',
        conf_threshold: this.confThreshold,
        mode: metadata?.mode || 'evaluate',
        timestamp: new Date().toISOString(),
        ...metadata,
      },
      per_iou: {},
      mAP_50_95: 0.0,
      mAP_50: 0.0,
      mAP_75: 0.0,
      mAP_mean: 0.0,
      is_coco_standard: false,
      image_count: this.imageCounter,
      total_predictions: this.allPredictions.length,
      total_ground_truths: this.allGroundTruths.length,
      per_class_summary: {},
    };

    const mapValues: number[] = [];

    for (const iouTh of this.iouThresholds) {
      const iouKey = `iou_${iouTh.toFixed(2)}`;
      const classMetrics: Record<string, ClassMetric> = {};
      const classAps: number[] = [];
      const classPrecs: number[] = [];
      const classRecs: number[] = [];

      for (const cls of allClasses) {
        const clsPreds = this.allPredictions.filter((p) => p.label === cls);
        const clsGts = this.allGroundTruths.filter((g) => g.label === cls);

        const [ap, precision, recall] = this.calculateAPForClass(clsPreds, clsGts, iouTh);

        classMetrics[cls] = {
          AP: ap,
          precision,
          recall,
          num_ground_truth: clsGts.length,
          num_predictions: clsPreds.length,
        };

        if (clsGts.length > 0) {
          classAps.push(ap);
          classPrecs.push(precision);
          classRecs.push(recall);
        }
      }

      const mAP = classAps.length > 0 ? classAps.reduce((a, b) => a + b, 0) / classAps.length : 0.0;
      const meanPrecision = classPrecs.length > 0 ? classPrecs.reduce((a, b) => a + b, 0) / classPrecs.length : 0.0;
      const meanRecall = classRecs.length > 0 ? classRecs.reduce((a, b) => a + b, 0) / classRecs.length : 0.0;

      report.per_iou[iouKey] = {
        mAP,
        mean_precision: meanPrecision,
        mean_recall: meanRecall,
        class_metrics: classMetrics,
      };

      mapValues.push(mAP);
      if (Math.abs(iouTh - 0.50) < 1e-4) report.mAP_50 = mAP;
      if (Math.abs(iouTh - 0.75) < 1e-4) report.mAP_75 = mAP;
    }

    report.mAP_mean = mapValues.length > 0 ? mapValues.reduce((a, b) => a + b, 0) / mapValues.length : 0.0;

    const cocoKeys = COCO_IOU_THRESHOLDS.map((th) => `iou_${th.toFixed(2)}`);
    const isCocoStd = cocoKeys.every((k) => k in report.per_iou);
    report.is_coco_standard = isCocoStd;

    if (isCocoStd) {
      const cocoMaps = cocoKeys.map((k) => report.per_iou[k].mAP);
      report.mAP_50_95 = cocoMaps.reduce((a, b) => a + b, 0) / cocoMaps.length;
    } else {
      report.mAP_50_95 = report.mAP_mean;
    }

    // Per-class summary across thresholds
    for (const cls of allClasses) {
      const clsApsAcross = this.iouThresholds.map((th) => report.per_iou[`iou_${th.toFixed(2)}`].class_metrics[cls]?.AP ?? 0.0);
      const ap50 = report.per_iou['iou_0.50']?.class_metrics[cls]?.AP ?? 0.0;
      const ap75 = report.per_iou['iou_0.75']?.class_metrics[cls]?.AP ?? 0.0;
      const prec50 = report.per_iou['iou_0.50']?.class_metrics[cls]?.precision ?? 0.0;
      const rec50 = report.per_iou['iou_0.50']?.class_metrics[cls]?.recall ?? 0.0;

      let apCoco = 0.0;
      if (isCocoStd) {
        const cocoClsAps = cocoKeys.map((k) => report.per_iou[k].class_metrics[cls]?.AP ?? 0.0);
        apCoco = cocoClsAps.reduce((a, b) => a + b, 0) / cocoClsAps.length;
      } else {
        apCoco = clsApsAcross.length > 0 ? clsApsAcross.reduce((a, b) => a + b, 0) / clsApsAcross.length : 0.0;
      }

      const numGts = this.allGroundTruths.filter((g) => g.label === cls).length;
      const numPreds = this.allPredictions.filter((p) => p.label === cls).length;

      report.per_class_summary[cls] = {
        AP_50_95: apCoco,
        AP_50: ap50,
        AP_75: ap75,
        AP_mean: clsApsAcross.length > 0 ? clsApsAcross.reduce((a, b) => a + b, 0) / clsApsAcross.length : 0.0,
        precision_50: prec50,
        recall_50: rec50,
        num_ground_truth: numGts,
        num_predictions: numPreds,
      };
    }

    return report;
  }
}
